issue #3126 Add support for multiple offline compose drafts

limonte opened the pull request:
This PR allows users to create several offline compose drafts. It also takes care of migrating the existed compose offline drafts, so nothing will be lost for users.

close #3126

<img width="617" alt="CleanShot 2021-09-30 at 11 20 41@2x" src="https://user-images.githubusercontent.com/6059356/135415198-077be682-e23c-46e1-b508-5c053fd5287b.png">

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities
- [ ] is documented clearly and usefully, or doesn't need documentation

limonte commented:
@tomholub the issue description wasn't specific about how to represent offline drafts, so I made a separate section "FlowCrypt offline drafts" with the list of drafts with their saving datetime as textual representation. Drafts are sorted by date desc.

Let me know if I missed something here or you'd like to change anything.

tomholub commented:
> @tomholub the issue description wasn't specific about how to represent offline drafts, so I made a separate section "FlowCrypt offline drafts" with the list of drafts with their localized datetime as textual representation. Drafts are sorted by date desc.
> 
> Let me know if I missed something here or you'd like to change anything.

Looks good. Haven't looked at code (will be afk) for the rest of the week - will let Roman review

rrrooommmaaa commented on extension/js/content_scripts/webmail/gmail-element-replacer.ts:
```diff
@@ -746,4 +749,47 @@ export class GmailElementReplacer implements WebmailElementReplacer {
+        $(this.sel.draftsList).append(offlineDraftsContainer); // xss-safe-value
+      }
+      if (Object.keys(offlineComposeDrafts).length) {
+        const draftsToRender: JQueryEl[] = [];
```
why is this array `draftsToRender` keyed differently than `storage.local_drafts`? Is this only to call `sort()` without a sorting function/lambda parameter?
`local-draft-compose-HMhqrZKsUo`  should be a good key, and its uniqueness is in line with the storage implementation. I see the click  handler operates with storage keys. 

The first thing that perplexed me is why `id` inside a draft object differs from the outer `draftId` in storage.
Is the only purpose to make it sortable? I guess this is unnecessary.
Also, it is named `id` as if its meaning is unknown (opaque) to the outside modules, and then on line 771 you apply the knowledge that it is actually a timestamp to format it with `new Date(...).toLocaleString()`.
This knowledge (requirement) should be explicitly reflected in the property name.
So I suggest for drafts to not have `id` field (or matching with the outside id, if required).
Have `timestamp` field instead. It is usually inique but not really required to be.
Do not create any dictionaries with `timestamp` as key. This is unnecessary complication and uniqueness requirement.

limonte replied:
Thank you for the advice @rrrooommmaaa, I followed it and the code is cleaner now 👍

rrrooommmaaa commented on test/source/tests/gmail.ts:
```diff
@@ -253,6 +253,19 @@ export const defineGmailTests = (testVariant: TestVariant, testWithBrowser: Test
+      const gmailPage = await openGmailPage(t, browser, '/');
+      // create compose draft
+      await gmailPage.waitAndClick('@action-secure-compose', { delay: 1 });
+      await createSecureDraft(t, browser, gmailPage, 'compose draft', { offline: true });
```
I would like to see adding more offline drafts here, checking the list of local drafts, opening one of them and checking that the correct one is opened.
Was there a problem with that?
I guess we can make the whole `gmailPage` offline for the period of testing? 
I managed to do this in a broswer, I guess, it can be puppeteered too.

limonte replied:
Done 👍

limonte commented:
Thank you for the review @rrrooommmaaa! I fixed the issues you've mentioned and also added one unrelated commit https://github.com/FlowCrypt/flowcrypt-browser/pull/4026/commits/4dfee3c6a66f1161b81f08f57414b1421aef0051 which moves one rule from tslint to eslint. The reason for that is that eslint would autofix that rule for us and tslint is deprecated.

limonte commented:
ping @rrrooommmaaa

rrrooommmaaa commented on conf/tslint.yaml:
```diff
@@ -10,7 +10,7 @@ rules: #
     arrow-parens: false
     no-console: false
-    prefer-const: true
+    prefer-const: false
```
why don't we prefer constants?

limonte replied:
I mentioned this change in [my previous comment](https://github.com/FlowCrypt/flowcrypt-browser/pull/4026#issuecomment-932503485):

> also added one unrelated commit 4dfee3c which moves the prefer-const rule from tslint to eslint. The reason for that is that eslint would autofix that rule for us and tslint is deprecated.

rrrooommmaaa commented on test/source/tests/gmail.ts:
```diff
@@ -253,6 +251,34 @@ export const defineGmailTests = (testVariant: TestVariant, testWithBrowser: Test
+      // compose draft 2 should be first in list as drafts are sorted by date descending
+      const draft = await pageHasSecureDraft(t, browser, urls[0], 'compose draft 2');
+      await Util.sleep(3); // the draft isn't being saved if start typing without this delay
+      await draft.type('@input-body', 'trigger saving to the clound', true);
```
"clound" typo

limonte replied:
fixed 👍

rrrooommmaaa reviewed (approved):
Looking much neater now, thank you!
